liblsp/Server.cpp

```diff
@@ -0,0 +1,391 @@
+		{"textDocument/references", bind(&Server::handle_textDocument_references, this, _1, _2)},
+	},
+	m_logger{std::move(_logger)}
+{
+}
+
+int Server::run()
+{
+	while (!m_exitRequested && !m_client.closed())
+	{
+		// TODO: receive() must return a variant<> to also return on <Transport::TimeoutEvent>,
+		// so that we can perform some idle tasks in the meantime, such as
+		// - lazy validation runs
+		// - check for results of asynchronous runs (in case we want to support threaded background jobs)
+		// Also, EOF should be noted properly as a <Transport::ClosedEvent>.
+		optional<Json::Value> const jsonMessage = m_client.receive();
+		if (jsonMessage.has_value())
+		{
+			try
+			{
+				handleMessage(*jsonMessage);
+			}
+			catch (std::exception const& e)
+			{
+				logError("Unhandled exception caught when handling message. "s + e.what());
+			}
+		}
+		else
+			logError("Could not read RPC request.");
+	}
+
+	if (m_shutdownRequested)
+		return EXIT_SUCCESS;
+	else
+		return EXIT_FAILURE;
+}
+
+void Server::invalidRequest(MessageId _id, string const& _methodName)
+{
+	// The LSP specification requires an invalid request to be respond with an InvalidRequest error response.
+	error(_id, ErrorCode::InvalidRequest, "Invalid request " + _methodName);
+}
+
+void Server::handleMessage(Json::Value const& _jsonMessage)
+{
+	string const methodName = _jsonMessage["method"].asString();
+
+	MessageId const id = _jsonMessage["id"].isInt()
+		? MessageId{_jsonMessage["id"].asInt()}
+		: _jsonMessage["id"].isString()
+			? MessageId{_jsonMessage["id"].asString()}
+			: MessageId{};
+
+	if (auto const handler = m_handlers.find(methodName); handler != m_handlers.end())
+	{
+		Json::Value const& jsonArgs = _jsonMessage["params"];
+		handler->second(id, jsonArgs);
+	}
+	else
+		error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+}
+
+void Server::handle_initializeRequest(MessageId _id, Json::Value const& _args)
+{
+	std::string rootUri;
+	if (Json::Value uri = _args["rootUri"]; uri)
+		rootUri = uri.asString();
+	else if (Json::Value rootPath = _args["rootPath"]; rootPath)
+		rootUri = "file://" + rootPath.asString();
+
+	Trace trace = Trace::Off; // The initial trace setting. If omitted trace is disabled ('off').  // TODO: ^^ should be overridable by the solls CLI param
+	if (Json::Value value = _args["trace"]; value)
+	{
+		string const name = value.asString();
+		if (name == "messages")
+			trace = Trace::Messages;
+		else if (name == "verbose")
+			trace = Trace::Verbose;
+		else if (name == "off")
+			trace = Trace::Off;
+	}
+
+	std::vector<WorkspaceFolder> workspaceFolders; // initial configured workspace folders
+	if (Json::Value folders = _args["workspaceFolders"]; folders)
+	{
+		for (Json::Value folder: folders)
+		{
+			WorkspaceFolder wsFolder{};
+			wsFolder.name = folder["name"].asString();
+			wsFolder.uri = folder["uri"].asString();
+			workspaceFolders.emplace_back(move(wsFolder));
```

Comment:
Does the following work?
```suggestion
			workspaceFolders.emplace_back(folder["name"].asString(), folder["uri"].asString());
```